packages/beacon-node/src/chain/archiveStore/differentialState/apply.ts

```diff
@@ -0,0 +1,95 @@
+import {PubkeyIndexMap} from "@chainsafe/pubkey-index-map";
+import {BeaconConfig} from "@lodestar/config";
+import {Logger} from "@lodestar/logger";
+import {IBeaconDb} from "../../../db/index.ts";
+import {IStateDiffCodec} from "../interface.ts";
+import {replayBlocks} from "../utils/replayBlocks.ts";
+import {StateRegenArtifacts} from "./fetch.ts";
+import {StateRegenPlan} from "./plan.ts";
+import {BeaconStateSnapshot} from "./ssz.ts";
+import {replayStateDifferentials} from "./stateDifferential.ts";
+import {beaconStateBytesToSnapshot, snapshotToBeaconStateBytes} from "./stateSnapshot.ts";
+
+export type StateRegenContext = {
+  codec: IStateDiffCodec;
+  config: BeaconConfig;
+  logger?: Logger;
+  pubkey2index: PubkeyIndexMap;
+  db: IBeaconDb;
+};
+
+export async function applyStateRegenPlan(
+  ctx: StateRegenContext,
+  plan: StateRegenPlan,
+  artifacts: StateRegenArtifacts
+): Promise<BeaconStateSnapshot> {
+  // When we start a node from a certain checkpoint which is usually
+  // not the snapshot epoch but we fetch it because of the fallback settings
+  if (plan.snapshotSlot !== artifacts.snapshot.slot) {
+    ctx.logger?.warn("Expected snapshot not found", {
+      expectedSnapshotSlot: plan.snapshotSlot,
+      availableSnapshotSlot: artifacts.snapshot.slot,
+    });
+  }
+
+  // TODO: Need to do further thinking if we fail here with fatal error
+  if (artifacts.missingDiffs.length) {
+    ctx.logger?.warn("Missing some diff states", {
+      snapshotSlot: plan.snapshotSlot,
+      diffPath: plan.diffSlots.join(","),
+      missingDiffs: artifacts.missingDiffs.join(","),
+    });
+  }
+
+  const orderedDiffs = [];
+  for (const diffSlot of plan.diffSlots) {
+    const diff = artifacts.diffs.find((d) => d.slot === diffSlot);
+    if (diff) {
+      orderedDiffs.push(diff);
+    }
+  }
+
+  if (orderedDiffs.length + artifacts.missingDiffs.length !== plan.diffSlots.length) {
+    throw new Error(`Can not find required state diffs ${plan.diffSlots.join(",")}`);
+  }
+
+  if (plan.blockReplay && orderedDiffs.at(-1)?.slot !== plan.blockReplay.fromSlot - 1) {
```

Comment:
```suggestion
  if (plan.blockReplay && orderedDiffs.at(0)?.slot !== plan.blockReplay.fromSlot - 1) {
```
`orderedDiffs` is sorted in ascending order, so shouldn't we check at 0 instead of -1?